styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -74,48 +76,79 @@
       .expireAfterWrite(CACHE_EXPIRY_DURATION.toMillis(), TimeUnit.MILLISECONDS)
       .build();
 
-  private static class CounterSnapshot {
+  private static class CounterSnapshotFactory {
 
-    private final String counterId;
-    private final Long limit;
-    private final Map<Integer, Long> shards;
+    private static final int TRANSACTION_GROUP_SIZE = 15;
 
-    public CounterSnapshot(Storage storage, String counterId) {
-      limit = getLimit(storage, counterId);
+    public static CounterSnapshot create(Storage storage, String counterId,
+                                         ExecutorService executorService) {
+
+      return new CounterSnapshot(storage, counterId,
+                                 getShards(storage, counterId, executorService));
+    }
 
+    private static Map<Integer, Long> getShards(Storage storage, String counterId,
+                                                ExecutorService executorService) {
       Map<Integer, Long> fetchedShards = storage.shardsForCounter(counterId);
       if (fetchedShards.size() < NUM_SHARDS) {
         // The counter probably has not been initialized (so we have empty QueryResults). Also
         // possible that a prior initialize() crashed halfway, or we got a partial list of shards in
         // QueryResults due to eventual consistency. In any case, repeated initialization eventually
         // creates all NUM_SHARDS shards.
-        initialize(storage, counterId);
+        initialize(storage, counterId, executorService);
         fetchedShards = storage.shardsForCounter(counterId);
       }
-      this.counterId = counterId;
-      this.shards = fetchedShards;
+      return fetchedShards;
     }
 
     /**
      * Idempotent initialization, so that we don't reset an existing shard to zero - counterId may
      * have already been initialized and incremented by another process.
      */
-    private static void initialize(Storage storage, String counterId) {
-      for (int i = 0; i < NUM_SHARDS; i++) {
-        final int shardIndex = i;
-        try {
-          storage.runInTransaction(tx -> {
-            final Optional<Shard> shard = tx.shard(counterId, shardIndex);
+    private static void initialize(Storage storage, String counterId,
+                                   ExecutorService executorService) {
+      for (int i = 0; i < NUM_SHARDS; i += TRANSACTION_GROUP_SIZE) {
```

Comment:
This double `for` loop seems buggy, with 128 shards and group size of 15 we end up creating 135 shards. This should be tested as well.

Reply:
You're right, there was one more condition to avoid this, which I've removed by mistake